File: Lista de Exercício/Golang/Calculadora.go
// Questão 10, 11, 12 e 13

package main

import (
	"fmt"
)

type Calculadora struct{}

func (c *Calculadora) SomarDois(a, b int) int {
	return a + b
}

func (c *Calculadora) SomarTres(a, b, c int) int {
	return a + b + c
}

type Funcionario interface {
	CalcularSalario() float64
	Nome() string
}

type FuncionarioHorista struct {
	nome             string
	horasTrabalhadas float64
	valorHora        float64
}

func (f *FuncionarioHorista) CalcularSalario() float64 {
	return f.horasTrabalhadas * f.valorHora
}

func (f *FuncionarioHorista) Nome() string {
	return f.nome
}

type FuncionarioAssalariado struct {
	nome          string
	salarioMensal float64
}

func (f *FuncionarioAssalariado) CalcularSalario() float64 {
	return f.salarioMensal
}

func (f *FuncionarioAssalariado) Nome() string {
	return f.nome
}

type Produto struct {
	Nome  string
	Preco float64
}

func SomarProdutos(p1, p2 Produto) Produto {
	return Produto{
		Nome:  fmt.Sprintf("%s + %s", p1.Nome, p2.Nome),
		Preco: p1.Preco + p2.Preco,
	}
}

type Matematica struct{}

func (m *Matematica) Fatorial(n int) (int, error) {
	if n < 0 {
		return 0, fmt.Errorf("o fatorial não está definido para números negativos")
	}
	if n == 0 || n == 1 {
		return 1, nil
	}
	result, err := m.Fatorial(n - 1)
	if err != nil {
		return 0, err
	}
	return n * result, nil
}

func main() {
	calc := &Calculadora{}
	fmt.Println(calc.SomarDois(5, 3))
	fmt.Println(calc.SomarTres(5, 3, 2))

	horista := &FuncionarioHorista{nome: "João", horasTrabalhadas: 160, valorHora: 20}
	assalariado := &FuncionarioAssalariado{nome: "Maria", salarioMensal: 3000}

	fmt.Printf("Salário de %s: R$%.2f\n", horista.Nome(), horista.CalcularSalario())
	fmt.Printf("Salário de %s: R$%.2f\n", assalariado.Nome(), assalariado.CalcularSalario())

	produto1 := Produto{Nome: "Camiseta", Preco: 50.0}
	produto2 := Produto{Nome: "Calça", Preco: 100.0}
	produtoSoma := SomarProdutos(produto1, produto2)

	fmt.Printf("Produto 1: %s: R$%.2f\n", produto1.Nome, produto1.Preco)
	fmt.Printf("Produto 2: %s: R$%.2f\n", produto2.Nome, produto2.Preco)
	fmt.Printf("Soma dos produtos: %s: R$%.2f\n", produtoSoma.Nome, produtoSoma.Preco)

	mat := &Matematica{}
	fatorial, err := mat.Fatorial(5)
	if err == nil {
		fmt.Printf("Fatorial de 5: %d\n", fatorial)
	} else {
		fmt.Printf("Erro ao calcular fatorial: %s\n", err)
	}

	fatorial, err = mat.Fatorial(0)
	if err == nil {
		fmt.Printf("Fatorial de 0: %d\n", fatorial)
	} else {
		fmt.Printf("Erro ao calcular fatorial: %s\n", err)
	}
}

File: Lista de Exercício/Java/Calculadora.java
// Questões 10, 11, 12 e 13 

package ListaDeExercicio.Java;

public class Calculadora {

    public static void main(String[] args) {
        // Testando a classe Calculadora
        Calculadora calc = new Calculadora();
        System.out.println(calc.somarDois(5, 3));
        System.out.println(calc.somarTres(5, 3, 2));

        // Criando instâncias de funcionários
        FuncionarioHorista horista = new FuncionarioHorista("João", 160, 20);
        FuncionarioAssalariado assalariado = new FuncionarioAssalariado("Maria", 3000);

        System.out.printf("Salário de %s: R$%.2f%n", horista.getNome(), horista.calcularSalario());
        System.out.printf("Salário de %s: R$%.2f%n", assalariado.getNome(), assalariado.calcularSalario());

        // Criando e somando produtos
        Produto produto1 = new Produto("Camiseta", 50.0);
        Produto produto2 = new Produto("Calça", 100.0);
        Produto produtoSoma = produto1.somar(produto2);

        System.out.println("Produto 1: " + produto1);
        System.out.println("Produto 2: " + produto2);
        System.out.println("Soma dos produtos: " + produtoSoma);

        // Testando o método estático de fatorial
        System.out.printf("Fatorial de 5: %d%n", Matematica.fatorial(5));
        System.out.printf("Fatorial de 0: %d%n", Matematica.fatorial(0));
    }

    public int somarDois(int a, int b) {
        return a + b;
    }

    public int somarTres(int a, int b, int c) {
        return a + b + c;
    }
}

// Classe abstrata Funcionario
abstract class Funcionario {
    protected String nome;

    public Funcionario(String nome) {
        this.nome = nome;
    }

    public String getNome() {
        return nome;
    }

    public abstract double calcularSalario();
}

// Classe FuncionarioHorista
class FuncionarioHorista extends Funcionario {
    private int horasTrabalhadas;
    private double valorHora;

    public FuncionarioHorista(String nome, int horasTrabalhadas, double valorHora) {
        super(nome);
        this.horasTrabalhadas = horasTrabalhadas;
        this.valorHora = valorHora;
    }

    @Override
    public double calcularSalario() {
        return horasTrabalhadas * valorHora;
    }
}

// Classe FuncionarioAssalariado
class FuncionarioAssalariado extends Funcionario {
    private double salarioMensal;

    public FuncionarioAssalariado(String nome, double salarioMensal) {
        super(nome);
        this.salarioMensal = salarioMensal;
    }

    @Override
    public double calcularSalario() {
        return salarioMensal;
    }
}

// Classe Produto
class Produto {
    private String nome;
    private double preco;

    public Produto(String nome, double preco) {
        this.nome = nome;
        this.preco = preco;
    }

    public Produto somar(Produto outro) {
        if (outro instanceof Produto) {
            String novoNome = this.nome + " + " + outro.nome;
            double novoPreco = this.preco + outro.preco;
            return new Produto(novoNome, novoPreco);
        } else {
            throw new IllegalArgumentException("Operação não suportada");
        }
    }

    @Override
    public String toString() {
        return String.format("%s: R$%.2f", nome, preco);
    }
}

// Classe Matematica com método estático para fatorial
class Matematica {
    public static int fatorial(int n) {
        if (n < 0) {
            throw new IllegalArgumentException("O fatorial não está definido para números negativos");
        }
        if (n == 0 || n == 1) {
            return 1;
        }
        return n * fatorial(n - 1);
    }
}
